routing/rt/rt.go

```diff
@@ -0,0 +1,324 @@
+// Copyright (c) 2016 Bitfury Group Limited
+// Distributed under the MIT software license, see the accompanying
+// file LICENSE or http://www.opensource.org/licenses/mit-license.php
+
+package rt
+
+import (
+	"encoding/gob"
+	"fmt"
+	"github.com/lightningnetwork/lnd/routing/rt/graph"
+	"io"
+)
+
+type OperationType uint32
+const (
+	ADD_CHANNEL OperationType = iota
+	REMOVE_CHANNEL
+)
+
+func (t OperationType) String()string {
+	switch t {
+	case ADD_CHANNEL:
+		return "<ADD_CHANNEL>"
+	case REMOVE_CHANNEL:
+		return "<REMOVE_CHANNEL>"
+	default:
+		return "<UNKNOWN_TYPE>"
+	}
+}
+type Channel struct {
+	Id1, Id2 graph.ID
+	EdgeID   graph.EdgeID
+	Info     *graph.ChannelInfo
+}
+
+type ChannelOperation struct {
+	Id1, Id2  graph.ID
+	EdgeID    graph.EdgeID
+	Info      *graph.ChannelInfo
+	Operation OperationType // One of ADD_CHANNEL, REMOVE_CHANNEL
+}
+
+func (c ChannelOperation) String () string {
+	return fmt.Sprintf(
+		"ChannelOperation[%v %v %v %v %v]",
+		c.Id1,
+		c.Id2,
+		c.EdgeID,
+		c.Info,
+		c.Operation,
+	)
+}
+
+func (op *ChannelOperation) Copy() ChannelOperation {
+	return ChannelOperation{
+		Id1:       op.Id1,
+		Id2:       op.Id2,
+		EdgeID:    op.EdgeID,
+		Info:      op.Info.Copy(),
+		Operation: op.Operation,
+	}
+}
+
+type DifferenceBuffer []ChannelOperation
+
+func (d DifferenceBuffer) String() string {
+	s := ""
+	for i:=0; i<len(d); i++ {
+		s += fmt.Sprintf("  %v\n", d[i])
+	}
+	s1 := ""
+	if len(d) > 0 {
+		s1 = "\n"
+	}
+	return fmt.Sprintf("DifferenceBuffer[%v%v]", s1, s)
+}
+
+func NewDifferenceBuffer() *DifferenceBuffer {
+	d := make(DifferenceBuffer, 0)
+	return &d
+}
+
+func (diffBuff *DifferenceBuffer) IsEmpty() bool {
+	return len(*diffBuff) == 0
+}
+
+func (diffBuff *DifferenceBuffer) Clear() {
+	*diffBuff = make([]ChannelOperation, 0)
+}
+
+func (diffBuff *DifferenceBuffer) Copy() *DifferenceBuffer {
+	d := make([]ChannelOperation, 0, len(*diffBuff))
+	for _, op := range *diffBuff {
+		d = append(d, op.Copy())
+	}
+	return (*DifferenceBuffer)(&d)
+}
+
+type RoutingTable struct {
+	// TODO: Add locking
+	G    *graph.Graph
+	diff []*DifferenceBuffer
+}
+
+func (rt *RoutingTable) ShortestPath(src, dst graph.ID) ([]graph.ID, error) {
+	return graph.ShortestPath(rt.G, src, dst)
+}
+
+func NewRoutingTable() *RoutingTable {
+	return &RoutingTable{
+		G:    graph.NewGraph(),
+		diff: make([]*DifferenceBuffer, 0),
+	}
+}
+
+func (rt *RoutingTable) AddNode(IDs ...graph.ID) {
+	for _, ID := range IDs {
+		rt.G.AddNode(ID)
+	}
+}
+
+func (rt *RoutingTable) AddChannel(Node1, Node2 graph.ID, edgeID graph.EdgeID, info *graph.ChannelInfo) {
+	rt.AddNode(Node1, Node2)
+	if !rt.HasChannel(Node1, Node2, edgeID) {
+		rt.G.AddUndirectedEdge(Node1, Node2, edgeID, info)
+		for _, diffBuff := range rt.diff {
+			*diffBuff = append(*diffBuff, ChannelOperation{Node1, Node2, edgeID, info, ADD_CHANNEL})
```

Comment:
Because different neighbors may have different pending updates. Between sending NeighborUPD and receiving NeighborACK some updates may occur. 
Consider scenario:
A has two neighbors B, C
1. A sends updates to B,C
2. B responds with NeighborACK, C does not
3. New update arrives to A. A can send it to B, but cannot send it to C(because NeighborACK is not received). After sending updates to B, difference buffers for B, C will be different.